### ISSUE TITLE
Unable to reproduce sha256/shaRound constraints number

### ISSUE THREAD

**Issue by collinzrj** [circify/circ#101]
According to Fig 19 of the CirC paper, the circuit size of sha256/shaRound in the ZoKrates standard library compiled by CirC should be 28949, but the circuit size I get with current version of CirC https://github.com/circify/circ/commit/4686854248485f59088898832bff78b3a184a0e2 is 33153

<img width="705" alt="Screen Shot 2022-09-07 at 8 35 34 PM" src="https://user-images.githubusercontent.com/44433088/189008170-5133f619-a235-4992-b9eb-b71038618689.png">

Commands I run
```
cargo build --release --features r1cs,smt,zok --example circ
./target/release/examples/circ ./third_party/ZoKrates/zokrates_stdlib/stdlib/hashes/sha256/shaRound.zok r1cs
```

Log I get
```
Options { path: "./third_party/ZoKrates/zokrates_stdlib/stdlib/hashes/sha256/shaRound.zok", frontend: FrontendOptions { language: Auto, value_threshold: None, rec_limit: 5, lint_prim_rec: false, z_isolate_asserts: false }, parties: 2, backend: R1cs { prover_key: "P", verifier_key: "V", lc_elimination_thresh: 50, action: Count } }
Done with IR optimization
Converting to r1cs
Pre-opt R1cs size: 52244
Final R1cs size: 33153
```

How can I reduce the circuit size of shaRound in current version of CirC?

**Comment by alex-ozdemir** [comment]
Hi Collin,

I'm excited that you're digging into CirC!

As described in Section 4.3, the paper's experiments are run using CirC "v1.0"
(the Haskell prototype). It's available here[1], but we don't develop or
maintain it. The two implementations have the same architecture, but do differ
in small ways. Some of those differences are likely the cause of this.

I can look more closely next week. If you want to do some investigation
yourself, compiling sub-functionalities of SHA with both systems is a good
approach.

An idle question: is this particular functionality (SHA2) important to you? I
had heard that y'all's plan was to use a non-circ SHA2 implementation.

Cheers,
Alex

[1]: https://github.com/circify/compiler

**Comment by alex-ozdemir** [comment]
Hi Collin,

Okay, I did have some time to look at this. I found two things.

First, there was a large compiler performance bug that we were actually aware
of, but had not yet fixed on master. It's now [fixed][1], substantially improving
compile times (thanks to Evan Laufer).

Second, there was a SHA2-relevant optimization that wasn't triggering on
shaRound.zok. Now, it [does trigger][2].

The result of this change is that CirC now compiles shaRound into 29057
constraints (down from 33153 before). This is still a *bit* worse than the
reported constraint count. Further investigation could probably yield
improvements. On the plus side, Rust CirC compiles about 100x faster than
Haskell CirC (CPU time).

Cheers,
Alex

[1]: https://github.com/circify/circ/commit/f6c37b505bf56368ee8b8ed3b37ba00ca9102bd8
[2]: https://github.com/circify/circ/commit/c1bac114865a80ef49e2bfee409f2f97d6f5a9b6
